File: backend/src/services/write-off.service.ts
import * as db from '../db';
import AppError from '../utils/AppError';
import type { PaginatedResult } from '../types';

const WRITE_OFF_SELECT = `
  SELECT wo.*,
         an.name      AS asset_name,
         at.image_url AS asset_image,
         at.brand, at.model, at.size,
         u.name       AS created_by_name
  FROM  write_off_orders wo
  JOIN  asset_types  at ON at.id = wo.asset_type_id
  JOIN  asset_names  an ON an.id = at.asset_name_id
  JOIN  users        u  ON u.id  = wo.created_by
`;

export async function listWriteOffs(
  clubId: string,
  { asset_type_id, source, from_date, to_date, page = 1, limit = 20 }: {
    asset_type_id?: string;
    source?: string;
    from_date?: string;
    to_date?: string;
    page?: number | string;
    limit?: number | string;
  }
): Promise<PaginatedResult<Record<string, unknown>>> {
  const offset = (Number(page) - 1) * Number(limit);
  const conditions = ['wo.club_id = $1'];
  const params: unknown[] = [clubId];

  if (asset_type_id) conditions.push(`wo.asset_type_id = $${params.push(asset_type_id)}`);
  if (source)        conditions.push(`wo.source = $${params.push(source)}`);
  if (from_date)     conditions.push(`wo.created_at >= $${params.push(from_date)}`);
  if (to_date)       conditions.push(`wo.created_at < $${params.push(to_date)}`);

  const where = conditions.join(' AND ');

  const [{ rows }, { rows: countRows }] = await Promise.all([
    db.query<Record<string, unknown>>(
      `${WRITE_OFF_SELECT} WHERE ${where} ORDER BY wo.created_at DESC
       LIMIT $${params.push(Number(limit))} OFFSET $${params.push(offset)}`,
      params
    ),
    db.query<{ count: string }>(`SELECT COUNT(*) FROM write_off_orders wo WHERE ${where}`, params.slice(0, -2)),
  ]);

  return { data: rows, total: parseInt(countRows[0].count), page: Number(page), limit: Number(limit) };
}

export async function getWriteOff(
  id: string,
  clubId: string
): Promise<Record<string, unknown>> {
  const { rows } = await db.query<Record<string, unknown>>(
    `${WRITE_OFF_SELECT} WHERE wo.id = $1 AND wo.club_id = $2`,
    [id, clubId]
  );
  if (!rows.length) throw new AppError('Write-off order not found', 404);
  return rows[0];
}

export async function createWriteOff(
  clubId: string,
  operatorId: string,
  { asset_type_id, quantity, reason, notes }: {
    asset_type_id?: string;
    quantity?: number | string;
    reason?: string;
    notes?: string;
  }
): Promise<Record<string, unknown>> {
  if (!asset_type_id) throw new AppError('asset_type_id is required', 400);
  if (!quantity || Number(quantity) < 1) throw new AppError('quantity must be at least 1', 400);

  const qty = Number(quantity);

  const client = await db.pool.connect();
  try {
    await client.query('BEGIN');

    // Verify asset type belongs to this club and get available batches FIFO
    const { rows: batchRows } = await client.query<{ id: string; available_quantity: number }>(
      `SELECT ab.id, ab.available_quantity
       FROM asset_batches ab
       JOIN asset_types at ON at.id = ab.asset_type_id
       WHERE ab.asset_type_id = $1 AND at.club_id = $2
         AND ab.available_quantity > 0 AND ab.status != 'retired'
       ORDER BY ab.purchase_date ASC NULLS LAST, ab.created_at ASC`,
      [asset_type_id, clubId]
    );

    if (!batchRows.length) throw new AppError('Asset not found or no available stock', 404);

    const totalAvail = batchRows.reduce((s, r) => s + Number(r.available_quantity), 0);
    if (qty > totalAvail) {
      throw new AppError(
        `Cannot write off ${qty} units; only ${totalAvail} available in stock`,
        409
      );
    }

    // Deduct FIFO across batches
    let remaining = qty;
    for (const batch of batchRows) {
      if (remaining <= 0) break;
      const deduct = Math.min(remaining, Number(batch.available_quantity));
      remaining -= deduct;
      const availBefore = Number(batch.available_quantity);

      await client.query(
        `UPDATE asset_batches
         SET available_quantity = available_quantity - $1,
             total_quantity     = total_quantity - $1,
             updated_at         = NOW()
         WHERE id = $2`,
        [deduct, batch.id]
      );
      await client.query(
        `INSERT INTO stock_movements
           (club_id, asset_batch_id, operator_id, type, quantity_delta, quantity_before, quantity_after, notes)
         VALUES ($1,$2,$3,'write_off',$4,$5,$6,$7)`,
        [clubId, batch.id, operatorId, -deduct, availBefore, availBefore - deduct, reason ?? 'Manual write-off']
      );
    }

    const { rows } = await client.query<Record<string, unknown>>(
      `INSERT INTO write_off_orders (club_id, asset_type_id, quantity, reason, source, created_by, notes)
       VALUES ($1,$2,$3,$4,'manual',$5,$6) RETURNING *`,
      [clubId, asset_type_id, qty, reason ?? null, operatorId, notes ?? null]
    );

    await client.query('COMMIT');
    return getWriteOff(rows[0].id as string, clubId);
  } catch (err) {
    await client.query('ROLLBACK');
    throw err;
  } finally {
    client.release();
  }
}

File: backend-dotnet/src/SportStock.Api/Services/WriteOffService.cs
using System.Linq.Expressions;
using Microsoft.EntityFrameworkCore;
using SportStock.Api.Data;
using SportStock.Api.Data.Entities;
using SportStock.Api.Data.Enums;
using SportStock.Api.Dtos.Common;
using SportStock.Api.Dtos.WriteOffs;
using SportStock.Api.Exceptions;

namespace SportStock.Api.Services;

// Ports backend/src/services/write-off.service.ts. Manual write-offs from
// /write-offs use FIFO batch deduction; the loan-driven write-offs created
// by LoanService.ConfirmReturn flow through a different code path and are
// not touched here.
internal sealed class WriteOffService(SportStockDbContext db) : IWriteOffService
{
    private static readonly HashSet<string> ValidSourceFilter = new(StringComparer.Ordinal)
    {
        "manual", "loan_return", "loan_lost",
    };

    private static readonly Expression<Func<WriteOffOrder, WriteOffResponse>> Projection = w => new WriteOffResponse
    {
        Id = w.Id,
        ClubId = w.ClubId,
        AssetTypeId = w.AssetTypeId,
        Quantity = w.Quantity,
        Reason = w.Reason,
        Source = w.Source,
        LoanItemId = w.LoanItemId,
        CreatedBy = w.CreatedBy,
        Notes = w.Notes,
        CreatedAt = w.CreatedAt,
        UpdatedAt = w.UpdatedAt,
        AssetName = w.AssetType.AssetName.Name,
        AssetImage = w.AssetType.ImageUrl,
        Brand = w.AssetType.Brand,
        Model = w.AssetType.Model,
        Size = w.AssetType.Size,
        CreatedByName = w.CreatedByNavigation.Name,
    };

    public async Task<PaginatedResult<WriteOffResponse>> ListAsync(
        Guid clubId, ListWriteOffsQuery query, CancellationToken ct = default)
    {
        if (query.Page < 1) query.Page = 1;
        if (query.Limit < 1) query.Limit = 20;

        WriteOffSource? sourceFilter = null;
        if (!string.IsNullOrWhiteSpace(query.Source))
        {
            if (!ValidSourceFilter.Contains(query.Source))
                throw new AppException(
                    "source must be one of: manual, loan_return, loan_lost", 400);
            sourceFilter = query.Source switch
            {
                "manual" => WriteOffSource.Manual,
                "loan_return" => WriteOffSource.LoanReturn,
                "loan_lost" => WriteOffSource.LoanLost,
                _ => null,
            };
        }

        IQueryable<WriteOffOrder> source = db.WriteOffOrders
            .IgnoreQueryFilters()
            .Where(w => w.ClubId == clubId);

        if (query.AssetTypeId is { } typeId)
            source = source.Where(w => w.AssetTypeId == typeId);
        if (sourceFilter is { } sf)
            source = source.Where(w => w.Source == sf);
        if (query.FromDate is { } from)
            source = source.Where(w => w.CreatedAt >= from);
        if (query.ToDate is { } to)
            source = source.Where(w => w.CreatedAt < to);

        var total = await source.CountAsync(ct);
        var data = await source
            .OrderByDescending(w => w.CreatedAt)
            .Skip((query.Page - 1) * query.Limit)
            .Take(query.Limit)
            .Select(Projection)
            .ToListAsync(ct);

        return new PaginatedResult<WriteOffResponse>
        {
            Data = data,
            Total = total,
            Page = query.Page,
            Limit = query.Limit,
        };
    }

    public async Task<WriteOffResponse> GetAsync(Guid id, Guid clubId, CancellationToken ct = default)
    {
        var row = await db.WriteOffOrders
            .IgnoreQueryFilters()
            .Where(w => w.Id == id && w.ClubId == clubId)
            .Select(Projection)
            .FirstOrDefaultAsync(ct);
        return row ?? throw new AppException("Write-off order not found", 404);
    }

    public async Task<WriteOffResponse> CreateAsync(
        Guid clubId, Guid operatorId, CreateWriteOffRequest req, CancellationToken ct = default)
    {
        if (req.AssetTypeId is null)
            throw new AppException("asset_type_id is required", 400);
        if (req.Quantity is null || req.Quantity < 1)
            throw new AppException("quantity must be at least 1", 400);

        var qty = req.Quantity.Value;
        var typeId = req.AssetTypeId.Value;

        await using var tx = await db.Database.BeginTransactionAsync(ct);

        // FIFO across batches: oldest purchase_date first, NULLs last.
        var batches = await (
            from b in db.AssetBatches.IgnoreQueryFilters()
            join at in db.AssetTypes.IgnoreQueryFilters() on b.AssetTypeId equals at.Id
            where b.AssetTypeId == typeId
                  && at.ClubId == clubId
                  && b.AvailableQuantity > 0
                  && b.Status != AssetStatus.Retired
            orderby b.PurchaseDate ?? DateOnly.MaxValue, b.CreatedAt
            select b
        ).ToListAsync(ct);

        if (batches.Count == 0)
            throw new AppException("Asset not found or no available stock", 404);

        var totalAvail = batches.Sum(b => b.AvailableQuantity);
        if (qty > totalAvail)
            throw new AppException(
                $"Cannot write off {qty} units; only {totalAvail} available in stock", 409);

        var remaining = qty;
        var movementNote = req.Reason ?? "Manual write-off";
        foreach (var batch in batches)
        {
            if (remaining <= 0) break;
            var deduct = Math.Min(remaining, batch.AvailableQuantity);
            remaining -= deduct;
            var availBefore = batch.AvailableQuantity;

            batch.AvailableQuantity -= deduct;
            batch.TotalQuantity -= deduct;
            batch.UpdatedAt = DateTime.UtcNow;

            db.StockMovements.Add(new StockMovement
            {
                Id = Guid.NewGuid(),
                ClubId = clubId,
                AssetBatchId = batch.Id,
                OperatorId = operatorId,
                Type = StockMovementType.WriteOff,
                QuantityDelta = -deduct,
                QuantityBefore = availBefore,
                QuantityAfter = availBefore - deduct,
                Notes = movementNote,
            });
        }

        var order = new WriteOffOrder
        {
            Id = Guid.NewGuid(),
            ClubId = clubId,
            AssetTypeId = typeId,
            Quantity = qty,
            Reason = req.Reason,
            Source = WriteOffSource.Manual,
            CreatedBy = operatorId,
            Notes = req.Notes,
        };
        db.WriteOffOrders.Add(order);

        await db.SaveChangesAsync(ct);
        await tx.CommitAsync(ct);

        return await GetAsync(order.Id, clubId, ct);
    }
}
